activerecord/test/cases/autosave_association_test.rb

```diff
@@ -168,16 +168,13 @@ def test_cyclic_autosaves_do_not_add_multiple_validations
   end
 
   private
-    def assert_no_difference_when_adding_callbacks_twice_for(model, association_name)
```

Comment:
This assertion logic had 2 flaws.

1. It was looking for ivars that are never defined. My guess is that this logic was made redundant when how autosave callbacks are defined changed at some point, but this was never updated.
2. The reflection used for defining `has_and_belongs_to_many` autosave callbacks isn't actually the final reflection. It's the through reflection from which the `has_many` is defined [at this point](https://github.com/rails/rails/blob/c1489a8ca4a9b91e21725f6a139e82e803c9eaa6/activerecord/lib/active_record/associations.rb#L2101). This became obvious when the the last test above (`test_should_not_add_the_same_callbacks_multiple_times_for_has_and_belongs_to_many `) began failing once I update this logic to use `assert_not_called` instead.